### PR TITLE
Fix bug in shape assignment

### PR DESCRIPTION
**Pull request by r03ert0**
## PR Summary
`shape` returns first the y dimension, then the x dimension. The code had the order inverted

## PR Checklist

- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [ ] New features are documented, with examples if plot related.
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [ ] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

**Comment by jklymak** [comment]
That *does* seem to be incorrect, thanks for the PR.  

Any chance you can add a test for this?  I don't understand why no tests were failing or no tests were "broken" by this fix.

**Comment by r03ert0** [comment]
I checked `test_streamplot.py` and the problem is that the vector field used has a square shape (100x100), so it doesn't matter if dimensions are swapped. Maybe I can just modify this test to have a rectangular shape and make it fail?

**Comment by QuLogic** [comment]
That seems like the simplest change, yes.

**Comment by jklymak** [comment]
Agreed - a symmetric test is not a very good test

**Comment by jklymak** [comment]
@r03ert0 Any interest in coming back to this?

File: lib/matplotlib/streamplot.py [hunk 1]
```diff
@@ -575,7 +575,7 @@ def _integrate_rk12(x0, y0, dmap, f, maxlength):
         dx2 = ds * 0.5 * (k1x + k2x)
         dy2 = ds * 0.5 * (k1y + k2y)
 
-        nx, ny = dmap.grid.shape
+        ny, nx = dmap.grid.shape
         # Error is normalized to the axes coordinates
         error = np.hypot((dx2 - dx1) / (nx - 1), (dy2 - dy1) / (ny - 1))
```
Comment (timhoffm) [inline lib/matplotlib/streamplot.py:580]: We could alternatively do
```suggestion
        error = np.hypot((dx2 - dx1) / (dmap.grid.nx - 1),
                         (dy2 - dy1) / (dmap.grid.ny - 1))
```
which avoids the error-prone shape order.

Comment (jklymak) [reply lib/matplotlib/streamplot.py:580]: .. a proper test should have caught the bug though.  I think its worth fixing the test.

Comment (r03ert0) [reply lib/matplotlib/streamplot.py:580]: hello @jklymak! I'm back in town. What would be required to move forward? Doing a PR for the test?

Comment (QuLogic) [reply lib/matplotlib/streamplot.py:580]: Just update the test in this PR.

**Comment by story645** [comment]
hi @r03ert0 do you have the time to come back to this?

**Comment by QuLogic** [comment]
I changed the test as so:
```diff
diff --git a/lib/matplotlib/tests/test_streamplot.py b/lib/matplotlib/tests/test_streamplot.py
index 3ed1118817..74cf4f4003 100644
--- a/lib/matplotlib/tests/test_streamplot.py
+++ b/lib/matplotlib/tests/test_streamplot.py
@@ -13,14 +13,14 @@ on_mac = (sys.platform == 'darwin')
 
 
 def velocity_field():
-    Y, X = np.mgrid[-3:3:100j, -3:3:100j]
+    Y, X = np.mgrid[-3:3:100j, -3:3:200j]
     U = -1 - X**2 + Y
     V = 1 + X - Y**2
     return X, Y, U, V
 
 
 def swirl_velocity_field():
-    x = np.linspace(-3., 3., 100)
+    x = np.linspace(-3., 3., 200)
     y = np.linspace(-3., 3., 100)
     X, Y = np.meshgrid(x, y)
     a = 0.1
@@ -70,8 +70,8 @@ def test_linewidth():
 def test_masks_and_nans():
     X, Y, U, V = velocity_field()
     mask = np.zeros(U.shape, dtype=bool)
-    mask[40:60, 40:60] = 1
-    U[:20, :20] = np.nan
+    mask[40:60, 80:120] = 1
+    U[:20, :40] = np.nan
     U = np.ma.array(U, mask=mask)
     # Compatibility for old test image
     ax = plt.figure().subplots()
```

The diff (for PNGs only) of _`master`_ from square to rectangular tests is here: https://github.com/QuLogic/matplotlib/commit/c2bf8b43ce60b85ec7c63c8a1858b53f1840bed9

The diff (for PNGs only) of _this branch_ (merged to master first) from square to rectangular tests is here: https://github.com/QuLogic/matplotlib/commit/19f22393fab5e562751ca9459aa7d58a56a85d16

And the diff between `master` with rectangular tests, and this branch with rectangular tests is here: https://github.com/QuLogic/matplotlib/compare/19619-rectangular..19619-fix-shape

As I mentioned on the call, all of them are different from each other, so I'm not certain which one is correct.

**Comment by tacaswell** [comment]
In summary:

 - the bug only shows up when the grid is un-evenly spaced (which is why the tests on the initial PR passed with no change)
 - the bug is in deciding if the gradient has changed too much from point to point (so we should no longer trust the first-order integration),  not all of the tests have vector fields with steep enough gradients to trigger this logic
 - The integrated flowlines changed because we are now sampling the underlying function at 2x in one direction which is giving us a better view of the underlying function so it makes sense that the integration will be change.
 - the integrated flow lines would only have been miss-leading in cases with wildly different x/y steps and extremely rapidly changing gradients

The images in QuLogic@19f2239 are the "most right".

We probably should also add an API change note to the effect of "computed flow lines may change for streamplot with un-even grids".

**Comment by QuLogic** [comment]
> We probably should also add an API change note to the effect of "computed flow lines may change for streamplot with un-even grids".

Should we still backport to 3.4.3 then, or stick to 3.5?

**Comment by tacaswell** [comment]
I think stick to v3.5.  Given that this is a very old bug, I do not think it is urgent to get in for 3.4.3 and given that it does change outputs waiting for 3.5 seems safer.

**Comment by QuLogic** [comment]
OK, if people are okay with those images, I will update the SVG and PDF (or possibly just increase the tolerance on the one or two that are just bit changes.)

**Comment by QuLogic** [comment]
Alright, I've updated all the formats now. For those where changes were not visible to a viewer, I only updated the tolerance. For those that were updated, I also dropped the back-compat code in the tests, so they are slightly different from what they were before.

**Review by tacaswell — APPROVED** [review]

**Review by jklymak — APPROVED** [review]

**Pull request merged**